### PR TITLE
Add constructor for Line from elements at locations

### PR DESCRIPTION
**Pull request by eltos**
## Description
Add a constructor for `Line` which allows passing a sequence definition as discussed below and inserts drift spaces as needed.  
This PR also introduces a `Marker` element, which is inserted at the beginning of each sub-sequence such that elements can be placed with respect to it using `from`.

Requires: https://github.com/xsuite/xobjects/pull/78  
Related docs update: https://github.com/xsuite/xsuite/pull/256

## Checklist

Mandatory: 

- [x] I have added tests to cover my changes
- [x] All the tests are passing, including my new ones
- [x] I described my changes in this PR description

Optional:

- [x] The code I wrote follows good style practices (see [PEP 8](https://peps.python.org/pep-0008/) and [PEP 20](https://peps.python.org/pep-0020/)).
- [x] I have updated the docs in relation to my changes, if applicable
- [ ] ~~I have tested also GPU contexts~~

-------------------

## Original PR description
Add a constructor for `Line` which allows passing elements by location and adds the required drifts so one doesn't have to do it manually

```
from_thin_elements_at(loc_elements, line_length, **kwargs)
    Constructs a line from thin elements at fixed locations, inserting drift spaces as needed
    
    Args:
        loc_elements: List of (loc, element) tuples with location and element.
        line_length: Total length of line to determine drift behind last element.
        kwargs: Arguments passed to constructor of the line
    
    Returns:
        An instance of Line
    
    Examples:
        >>> line = xt.Line.from_thin_elements_at((
                    (0.7, xt.Multipole(length=0.3, knl=[0, +0.50], ksl=[0, 0])),
                    (1.4, xt.Multipole(length=0.5, knl=[np.pi / 12], hxl=[np.pi / 12])),
                    (1.8, xt.Multipole(length=0.2, knl=[0, 0, 0.1])),
                    (2.1, xt.Multipole(length=0.3, knl=[0, -0.50], ksl=[0, 0])),
                    (2.8, xt.Multipole(length=0.5, knl=[np.pi / 12], hxl=[np.pi / 12])),
                ), line_length = 5)
```

### CODE REVIEW

**Comment by giadarol** [comment]
Thank you very much! We were thinking of something along these lines since a while.

We would propose a moregeneral API to allow optionally specifying the name of the elements or the positions with respect to other elements. We were thinking of something along these lines:

```python
elements={
 'q1': xt.Multipole(length=0.3, knl=[0, +0.50], ksl=[0, 0])),
 'd1': xt.Multipole(length=0.5, knl=[np.pi / 12], hxl=[np.pi / 12])))
...
}
line = xt.Line.from_sequence(elements = elements, length=5, 
    locations=[
        Location(s=0.7, name='q1'),
        Location(s=0.1, name='d1',from='q1')
        ....]
)
```
This follows the MAD-X naming convention where a "line" contains the dift, while a "sequence" is an association name/s and the drift are autogenerated.

In the future this could work also for thick elements. For now we could check that all the provided elements are thin and raise a NotImplementedError otherwise.

What do you think?

**Comment by eltos** [comment]
It's actually rather simple to do, if you want, I can include it in the PR:

```python
def from_sequence(sequence):    
    element_names = []
    element_locations = []
    elements = []
    for name, args in sequence.items():
        from_ = args.get('from', args.get('from_', 0))
        if isinstance(from_, str):
            from_ = element_locations[element_names.index(from_)]
        at = args.get('at', np.max(element_locations, initial=0))

        if 'seq' in args:
            # nested sequence
            element_names.append(name)
            elements.append(xt.Drift()) # TODO: here we need a marker element, for now use a zero length drift
            element_locations.append(from_ + at)
            # add elements recursively
            nam, loc, els = from_sequence(args['seq'])
            element_names.extend([f'{name}_{n}' for n in nam])
            element_locations.extend([from_ + at + s for s in loc])
            elements.extend(els)
            
        elif 'el' in args:
            # element definition
            element_names.append(name)
            elements.append(args['el'].copy()) # copy to allow re-using elements
            element_locations.append(from_ + at)
            
            
    return element_names, element_locations, elements

## Example

dipol = xt.Multipole(length=0.5, knl=[np.pi / 12], hxl=[np.pi / 12])
section = {
 'q1': dict(el=xt.Multipole(length=0.3, knl=[0, +0.50]), at=0.7),
 'd1': dict(el=dipol, at=0.1, from_='q1'),    
 'x1': dict(el=xt.LimitRect(), at=1.2),
}
sequence = {
 's1': dict(seq=section, at=0),
 's2': dict(seq=section, at=2, from_='s1'),
 'q1': dict(el=xt.Multipole(length=0.3, knl=[0, 0, 0.50]), at=5),
#...
}

print(f'name    s     element')
for a,b,c in zip(*from_sequence(sequence)):
    print(f'{a:6s}  {b:.2f}  {c}')
```

which gives:
```
name    s     element
s1      0.00  <xtrack.beam_elements.elements.Drift object at 0x7ff8ba34cee0>
s1_q1   0.70  <xtrack.beam_elements.elements.Multipole object at 0x7ff8ba27ba60>
s1_d1   0.80  <xtrack.beam_elements.elements.Multipole object at 0x7ff8ba27bdf0>
s1_x1   1.20  <xtrack.beam_elements.apertures.LimitRect object at 0x7ff8ba27be80>
s2      2.00  <xtrack.beam_elements.elements.Drift object at 0x7ff8ba27bd90>
s2_q1   2.70  <xtrack.beam_elements.elements.Multipole object at 0x7ff8ba27beb0>
s2_d1   2.80  <xtrack.beam_elements.elements.Multipole object at 0x7ff8ba27bdc0>
s2_x1   3.20  <xtrack.beam_elements.apertures.LimitRect object at 0x7ff8ba246dc0>
q1      5.00  <xtrack.beam_elements.elements.Multipole object at 0x7ff8ba27be20>
```

and then use the `from_thin_elements_at` method to add the drifts:
```python
def line_from_sequence(sequence, length):
    nam, loc, els = from_sequence(sequence)
    return xt.Line.from_thin_elements_at(zip(loc, els), line_length=length, element_names=nam, **kwargs)
```

Support for thick elements can later be added by adjusting drifts in `from_thin_elements_at` and by adding support for the `REFER` and `REFPOS` keywords in from_sequence.

**Comment by giadarol** [comment]
Indeed I like the idea of keeping the possibility of defining line with a minimal amount of typing. Nevertheless I would like to keep the possibility of using a more explicit definition.

I would propose the following API allowing 3 different degrees of verbosity:

 1.  Most explicit way as defined before (allows repetition of elements, which can be useful for very large machines and for 
```python
elements with large memory footprint):
elements={
 'q1': xt.Multipole(length=0.3, knl=[0, +0.50], ksl=[0, 0])),
 'd1': xt.Multipole(length=0.5, knl=[np.pi / 12], hxl=[np.pi / 12])))
...
}
line = xt.Line.from_sequence(elements = elements, length=5, 
    locations=[
        Location(s=0.7, name='q1'),
        Location(s=0.1, name='d1',from='q1')
        ....]
)
```

2. Elements defined on the fly (no dictionary), nemes are optional
```python
line = xt.Line.from_sequence(
    locations=[
        Location(s=0.7, element=xt.Multipole(length=0.3, knl=[0, +0.50], ksl=[0, 0])), name='q1'),
        Location(s=0.1, element=xt.Multipole(length=0.5, knl=[np.pi / 12], hxl=[np.pi / 12]))), name='d1', from='q1'),
        ....], 
    length=5, 

)
```

3. Compact definition through list of tuples (names are optional)
```python
line = xt.Line.from_sequence([
        (0.7, xt.Multipole(length=0.3, knl=[0, +0.50], ksl=[0, 0]), 'q1'),
        (0.8, xt.Multipole(length=0.3, knl=[0, +0.50], ksl=[0, 0])),
        ....], 
    length=5, 
)
```

The implementation should be easy:
 - in case 3. one builds location objects within the method and falls back on case 2.
 - in case 2 one builds the dictionary and the missing name and falls back on case 1.

What do you think?

**Comment by eltos** [comment]
Maybe one can unify our ideas by having a common "what" parameter accepting elements and sub-sequences either by name or object; by allowing to pass the name separately (to prevent conflicts of equal names when repeating elements); by allowing for nested sub-sequence (which will be useful to build a line by means of equal sections into which additional elements are placed); and maybe by renaming `Location` to `At` to reduce typing overhead such that there is no need for a separate list-of-tuple method. Something like

```python
class At:
    def __init__(self, s, what, *, from_=0, name=None):
        """
        Args:
            s (float): Location (in m) of what relative to from_
            what (str, BeamElement or list): Object to place here. Can be an instance of a BeamElement, 
                another sequence given as list of At, or the name of a named element or sequence.
            from_ (float or str, optional): Reference location for placement, can be the s coordinate (in m) 
                or the name of a previous element or sequence.
            name (str, optional): Name of the element to place here. If None, a name is chosen automatically.
        """
        self.s = s
        self.from_ = from_
        self.what = what
        self.name = name

def from_sequence(locations, length, *, elements=None, sequences=None, copy_elements=False):
    """
    Args:
        locations (list of At): Sequence definition.
        length: Total length (in m) of line. Determines drift behind last element.
        elements: Dictionary with named elements, which can be refered to by name in the sequence definion.
        sequences: Dictionary with named sub-sequences, which can be refered to by name in the sequence definion.
        copy_elements (bool): Whether to make copies of elements or not. By default, named elements are re-used
            which is memory efficient but does not allow to change parameters individually.
    """
    for at in locations:
        if isinstance(at.what, BeamElement) or at.what in elements:
            ... # place element
        else:
            sequence = sequences[at.what] if isinstance(at.what, str) else at.what
            ... # place sub-sequence by recursion

```

Which would allow all of the above:
```python
from xtrack import Line, At, Multipole, ParticlesMonitor
elements = {
   'quad': Multipole(length=0.3, knl=[0, +0.50], ksl=[0, 0])),
   'bend': Multipole(length=0.5, knl=[np.pi / 12], hxl=[np.pi / 12])))
    ...
}
sequences = dict(
  section = [At(...), At(...)],
)
monitor = xt.ParticlesMonitor(...)
line = Line.from_sequence(length=5, elements = elements, sequences=sequences, 
    locations=[

        # by name, allowing repetition
        At(0.7, 'quad'),
        At(0.1, 'bend', from_='quad'),
        At(1.7, 'quad', name='quad2'), # repeat element 'quad', but give it a unique name
        At(1.8, 'bend'), # the implementation can also append a suffix automatically to ensure unique names
        ....

        # on the fly, most compact
        At(0.3, Multipole(...)),  # overhead to my initial proposal is negligible
        At(0.7, Multipole(...), name='quad3'),  # but it still allows to assign names as desired
        At(0.1, Multipole(...), name='bend3', from_='quad2'),
        ....

        # nested sequence (by name or on the fly)
        At(0, "section", name="section_1"),
        At(0.1, monitor, from_="section_1"),
        At(0, [At(...), At(...)]),
    ]
)
```

What do you think?

**Comment by giadarol** [comment]
I like it. Could you please update the pull request accordingly?

Only one thing: I find "At" a bit misleading, and it's not that I like "Location" much better though. Should we stick to the MAD-X naming and call it "Node"?

**Comment by eltos** [comment]
Okay, I will update the PR. "Node" is perfect.

**Comment by eltos** [comment]
@giadarol I have updated the PR and it's description. Please have a look.

File: xtrack/line.py [hunk 1]
```diff
@@ -506,6 +689,20 @@ def get_s_position(self, at_elements=None, mode="upstream"):
         else:
             return s
 
+    def remove_marker(self, inplace=True):
```
Comment (giadarol) [inline xtrack/line.py:692]: Rename to remove_markers

**Review by giadarol — APPROVED** [review]
Great work! Thanks a lot!!!

**Pull request merged**